lib/Sema/NameBinding.cpp

```diff
@@ -36,60 +37,539 @@
+    /// the other member variables.
+    NullablePtr<ImportDecl> ID;
+
+    SourceLoc importLoc;
+    ImportOptions options;
+    StringRef privateImportFileName;
+    Located<ImportKind> importKind;
+    ModuleDecl::AccessPathTy modulePath;
+    ModuleDecl::AccessPathTy declPath;
+
+    NullablePtr<DeclAttributes> attrs;
+
+    /// Create an UnboundImport for a user-written import declaration.
+    explicit UnboundImport(ImportDecl *ID);
+
+    /// Make sure the import is not a self-import.
+    bool checkNotTautological(const SourceFile &SF);
+
+    /// Make sure the module actually loaded, and diagnose if it didn't.
+    bool checkModuleLoaded(ModuleDecl *M, SourceFile &SF);
+
+    /// Find the top-level module for this module. If \p M is not a submodule,
+    /// returns \p M. If it is a submodule, returns either the parent module of
+    /// \p M or \c nullptr if the current module is the parent (which can happen
+    /// in a mixed-language framework).
+    NullablePtr<ModuleDecl> getTopLevelModule(ModuleDecl *M, SourceFile &SF);
+
+    /// Diagnose any errors concerning the \c @_exported, \c @_implementationOnly,
+    /// \c @testable, or \c @_private attributes, including a
+    /// non-implementation-only import of a fragile library from a resilient one.
+    void validateOptions(NullablePtr<ModuleDecl> topLevelModule, SourceFile &SF);
+
+    /// Create an \c ImportedModuleDesc from the information in this
+    /// UnboundImport.
+    ImportedModuleDesc makeDesc(ModuleDecl *module) const {
+      return ImportedModuleDesc({ declPath, module }, options,
+                                privateImportFileName);
+    }
 
-namespace {  
-  class NameBinder {    
-  public:
+  private:
+    void validatePrivate(ModuleDecl *topLevelModule);
+    void validateImplementationOnly(ASTContext &ctx);
+    void validateTestable(ModuleDecl *topLevelModule);
+    void validateResilience(NullablePtr<ModuleDecl> topLevelModule,
+                            SourceFile &SF);
+
+    /// Diagnoses an inability to import \p modulePath in this situation and, if
+    /// \p attrs is provided and has an \p attrKind, invalidates the attribute and
+    /// offers a fix-it to remove it.
+    void diagnoseInvalidAttr(DeclAttrKind attrKind, DiagnosticEngine &diags,
+                             Diag<Identifier> diagID);
+  };
+
+  /// Represents an import whose options have been checked and module has been
+  /// loaded, but its scope (if any) has not been validated.
+  struct BoundImport {
+    UnboundImport unbound;
+    ImportedModuleDesc desc;
+    ModuleDecl *module;
+    bool needsScopeValidation = false;
+
+    BoundImport(UnboundImport unbound, ImportedModuleDesc desc,
+                ModuleDecl *module, bool needsScopeValidation);
+
+    /// Validate the scope of the import, if needed.
+    void validateScope(SourceFile &SF);
+  };
+
+  class NameBinder : public DeclVisitor<NameBinder> {
```

Comment:
Mark `NameBinder` as final? I don't think anything is inheriting from it.

Reply:
Do you think it would be useful to have a small explanation for the different stages imports go through? Something like (assuming I've understood the details correctly):

```
[Note: import-name-binding-stages]
Imports are roughly processed in three stages:

1. Unbound: The import has been synthesized but is not known which ModuleDecl it corresponds to. 
2. Bound: We know which ModuleDecl the import corresponds to, but we may not have validated it fully.
3. Bound and validated: The import is bound to a ModuleDecl and has been fully validated so we are free to use it.
```

This explanation can be referred to in the doc comments for `addImport` and `bindImport` (right now, their doc comments seems to be overly focused on describing the implementation, which seems dissatisfying) with a `see [Note: import-name-binding-stages]`.

[Feel free to skip this suggestion if you feel like rephrasing it correctly might take too much time.]